### PR TITLE
Feature: Share links

### PR DESCRIPTION
**Pull request by shamoon**
## Proposed change

This PR implements "share links" aka publicly-accessible URLs that can be used to provide access to a doc. Feature image available at `feature-share-links` (warning: includes migrations)

Open to any feedback of course. Hopefully I havent missed anything (big).

- Links are of the format `{paperless-url}/share/{randomly-generated-slug}`
- Dropdown menu added to doc details, consolidated 'redo ocr' + 'more like this' under 'actions'
- Sane defaults for expiration = 7 days (other options: 1, 30, never). Can also choose to share 'archive' or 'original'
- To simplify things there's no edit UI for links, just delete and make a new one
- Redirects to login with an error message if link expired or not found

https://github.com/paperless-ngx/paperless-ngx/assets/4887959/31234f54-9f23-45f7-814d-08055200444a

Screenshot 2023-08-14 at 1 51 20 PM
Screenshot 2023-08-14 at 1 51 26 PM

Screenshot 2023-08-14 at 11 29 43 AM
Screenshot 2023-08-14 at 11 29 34 AM

Fixes #375 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain):

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by stumpylog** [comment]
Great feature, I think people will like it a lot.  codecov seems to think there's some lines that aren't getting hit still, maybe some more testing for the add/remove permissions especially?

**Comment by shamoon** [comment]
Ok lets see if codecov happier with this

**Comment by danieldietzler** [comment]
You've probably already considered them but here are some ideas I'd like to propose:

- Per file password support: for highly confidential documents it would be sensible to have this (optional) extra layer of security. (Personal note: I try to keep my firewall settings as tight as possible, so I plan on only exposing /share to the open internet (probably applies to more people than just me). This means that it would be helpful if the password form were also available at /share and not /account for example.) 
- Having the ability to extend the link validity in the share menu
- I don't know how sensible this is because I have now idea how to NOT make this clustered and ugly, but having some customisability concerning the time durations would be nice. The best case herby obviously would be to be able to specify any duration in any unit (dropdown for Minutes/Hours/Days/Months(?)) and a number input. This is because I can imagine (again for confidential documents) that you would want the link to be available for e.g. just an hour, just so that the recipient can get the document. I know that a link can be revoked manually but I wouldn't want to trust myself that I remember to revoke it an hour later.
- What happens to expired links? Are they just gone from the share menu, or are they displayed as expired? In case there would be the ability to extend links, expired links should probably be able to be extended (refreshed) as well (maybe up to like 30 days after which they get deleted?)

**Comment by smseidl** [comment]
@shamoon Outside of the expiration time on the links, how is this different that using `https://paperless.XXXXXXX.com/documents/<id>`?  I currently click on the open document link to get a link to paste into a program I am writing (Medical/EOB Tracking).
image

~~When I first saw this, I thought it might be something useful, but the links expiring doesn't help me.~~  After watching the video it seems like to does more than I originally thought.  I will try to test it out some time in the next 2 weeks.

Would it be possible to have an icon on the open file area to auto-copy a default shareable link to save mouse clicks?
image

**Comment by shamoon** [comment]
Seems like you sorta answered your question, I missed the previous comment too. This does remind me I need to add an update to the docs to this.

- Expired links are just hidden once expired. My feeling is it's so easy to create a new one it's not really worth having the ability to fiddle with extending. And frankly from a UI standpoint I wanted to keep it dead-simple. Link expires in a few days but you need longer? Create a new one.
- In terms of expiration options, Im sure this is goldilocks kinda thing (aka no list is perfect for everyone). I dont particularly want to integrate a date-picker or even worse, date validation 😅. Again, goal is to just keep it simple. And I dont see a big need for 1 h vs 1 day but open to it.
- As for a password field it sounds... a bit messy. Unfortunately I think that'd have to be for a future enhancement to this.
- I dont think a link to this belongs in the sidebar. UI-wise it sounds odd (like not expected by the user) and complicated. E.g. the share link dropdown has a whole little UI unto itself and that doesnt seem appropriate to have pop out of the sidebar imho.

I did add some basic documentation of this.

**Comment by danieldietzler** [comment]
> Link expires in a few days but you need longer? Create a new one.

I get your idea but I'd argue "Create a new one" isn't always that simple. If I want to share documents from a business perspective, maybe with a lawyer or so, it will be pretty annoying for them when they have to wait for a new link and then maybe update some bookmarks internally, ...
In a private environment it is no issue, sure. I can just them text them the new link and they are good.
The only workaround for this would then be a custom URL, that always points a correct link. Sounds not ideal to me, although I really get the idea of keeping it as simple as possible.

**Review by stumpylog — CHANGES_REQUESTED** [review]
Definitely something I think people will like.  Just a few questions and some clean ups

**Comment by shamoon** [comment]
Thanks! I'll try to work through the suggestions, some more python learning for me =)

But to answer one thing, yes "never" expiring links is one of the default options and thats when you dont pass an expiration. Technically the API lets you pass any date (because why not?) but the UI just gives a few options.

Screenshot 2023-09-11 at 5 18 24 PM

File: src/documents/tests/test_views.py [hunk 1]
```diff
@@ -74,3 +82,70 @@ def test_index(self):
...
+            - Invalid request for share link is made
+            - Request for expired share link is made
+        THEN:
+            - Document is returned without need for login
+            - User is redirected to login with error
+            - User is redirected to login with error
+        """
+        dirs = setup_directories()
+
+        _, filename = tempfile.mkstemp(dir=dirs.originals_dir)
```
[the remark this answers is not in the document]
Comment (shamoon) [reply src/documents/tests/test_views.py:85]: Let me know if I did this correctly

File: src/documents/tests/test_api.py [hunk 2]
```diff
@@ -2558,6 +2560,119 @@ def test_tag_unique_name_and_owner_enforced_on_update(self):
         )
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
 
+    def test_create_share_links(self):
```
[the remark this answers is not in the document]
Comment (shamoon) [reply src/documents/tests/test_api.py:2563]: I'll be honest, didnt totally understand what you mean by this, like where to add the tzinfo, in https://github.com/paperless-ngx/paperless-ngx/blob/06d0037b8ed7c4b611fc2b8fb8bb054dbd2dd53a/src/documents/tests/test_api.py#L2593 ?

[the remark this answers is not in the document]
Comment (stumpylog) [reply src/documents/tests/test_api.py:2563]: Ok, 35428e759b33fac387eb396b6b45b7ad9a5ac997 fixes this.  It was the naive datetime there, comparing against the UTC aware datetime from the database

**Comment by shamoon** [comment]
Thanks for the fixes stumpy! all that staring at the screen and it was right there ha 🥵

**Comment by stumpylog** [comment]
Actually trying the image out now, one thing I see is when creating the link, it doesn't populate under the dropdown until I refresh.  So select an expiry, create, no link appears.  Click dropdown again to toggle it appearing again, no link.  Refresh, link is there.  Anything your wizardry in the front can do there?

**Comment by shamoon** [comment]
Oh interesting, that shouldnt be the case (e.g. if you watch the video above), let me look

**Comment by shamoon** [comment]
Weird, I can't reproduce that issue, see video:

https://github.com/paperless-ngx/paperless-ngx/assets/4887959/31cb5593-ffe7-4ff6-8b8e-b1db53f7d82e

(I tried in the feature image too)

Can you try emptying browser cache?

**Comment by stumpylog** [comment]
It seems consistent across private windows, Firefox and Chrome, after cleaning data for the site.  I see the POST to create, should there be another get request?

**Comment by stumpylog** [comment]
Curious, I only see POSTs:
image

A delete does POST/DELETE:
image

Most odd

**Comment by shamoon** [comment]
I have an idea! Is the document youre on id = 0?

**Comment by shamoon** [comment]
Oh never mind I see the screenshot, it's 1. Ugh

**Comment by shamoon** [comment]
Ok @stumpylog please give the new build a try when you have a sec. Glad to have caught the issue, was missed in testing because I was mocking the clipboard functions.

TLDR is javascript copying is only allowed in a secure context (I didnt get this because was testing over localhost and my online one does use https), so we need a fallback. Short of writing my own opted for a popular small utility package.

Let me know if this solves things for you!

**Review by stumpylog — APPROVED** [review]
Yep, that looks good now.  Links update, links work when not logged in, and stop working after deleted.  I didn't try to wait a day to check expiration, but the backend for that should be good
